### PR TITLE
Updates from PyCharm, April 2017

### PR DESCRIPTION
**Pull request by Elizaveta239**
Frame evaluation debugger for Python 3.6 and some other fixes.

### CODE REVIEW

**Comment by fabioz** [comment]
I started to review this and do some tests... I think there's a critical issue lying around:

Let's say you have the following code:

```
import time
start_time = time.time()

def method():
    a = 1
    a += 1

# do some busy work
print("Started main task")
x = 0
for i in range(40000000):
    x += 1
    method()
print("Completed main task")

print(('TotalTime>>%s<<' % (time.time()-start_time,)))
```

if you start the debugger without any breakpoints, wait for the `Started main task` message to be printed and then try to add a breakpoint in method(), it won't be triggered.

Also, there are some misconceptions there... I see you've created an InternalSetTracingThread() command which will never execute (it won't execute because it's bound to a thread_id from a thread which is not stopped (i.e.: it will never call `process_internal_commands` and thus will never have a chance to execute -- the only commands which can be bound to a thread_id are the ones which happen when a thread is stopped in do `do_wait_suspend`)).

You can probably use cython to access the internals of CPython to set the tracing from outside the current thread through some hackery (you can refer to attach.cpp, which is used in the attach to process for that) -- as you'll have the GIL held for that it should be straightforward... Ideally you should do unit-tests for this so that we make sure it works on the CI on the supported platforms (you can probably extract the related code from attach.cpp to some .hpp and use it on both places so that it's always up to date).

I also noted that you changed some of the tests which actually test this (i.e.: WriterThreadCase3 should not be changed -- https://github.com/fabioz/PyDev.Debugger/pull/82/files#diff-3e10f65360eccea414240b90f620e434R795)

Also, tests still seem to be failing on the ci.

**Comment by Elizaveta239** [comment]
Hi @fabioz! In your example the breakpoint inside `method()` will be hit, because we process internal commands inside frame evaluation function: https://github.com/Elizaveta239/PyDev.Debugger/blob/4545ed2f95ae3644ec0f4491c8c41598fa2d74a0/_pydevd_frame_eval/pydevd_frame_evaluator.pyx#L111
And this `InternalSetTracingThread` command will be executed when entering the new frame.
But you're right that this function won't work if the program doesn't enter the new frame and if it isn't suspend at any other breakpoint, debugger will skip such breakpoint. 
Thank you for the hint with attaching the process, you're right, we should use it in such cases.

Yes, I'm going to fix tests.

**Comment by fabioz** [comment]
I was thinking a bit more about it... maybe you could have a dummy tracing always set -- something as 

```
def trace_dispatch(frame, event, arg):
    return None

```
The overhead should be minimal in this case (you can probably do a cdef on Cython to make it even faster) and you wouldn't need to do any changes (just changing frame.f_trace would always work). It would simplify things quite a bit because you wouldn't need to keep any state to enable/disable the old debugging (and I believe should make it more stable as state in the debugger can change very fast).

Regarding the issue at hand, I find it strange that it never hits the breakpoint in the example I gave (and it's constantly entering/exiting the `method` function).

**Comment by Elizaveta239** [comment]
Yes, it's interesting idea, I would like to make some performance measurements.

**Comment by Elizaveta239** [comment]
Debugging with dummy tracing becomes almost 2 times slower (in the worst case). But debugger becomes much more stable, so, I think, I should implement it. And also we can remove all this stuff with `InternalSetTracingThread` command. 
Thank you very much!

**Comment by fitermay** [comment]
Maybe it's possible to do this more  by setting frame->f_trace on the last thread of every frame?

**Comment by Elizaveta239** [comment]
Hi! Thank you very for the review! Unfortunately, I won't be available during the next 3 days, but after that I'll answer to the questions and I'll make fixes according to your comments.

File: pydevd.py [hunk 1]
```diff
@@ -256,6 +258,9 @@ def __init__(self):
         self.show_return_values = False
         self.remove_return_values_flag = False
 
+        # this dict is used for frame evaluation, it holds thread_id -> set of frames
+        self.disable_tracing_after_exit_frames = {}
```
Comment (Elizaveta239) [inline pydevd.py]: Yes, but we still need a flag for disabling frame evaluation (if, for example, breakpoint was added during running). So, I've added a field "do_not_use_frame_eval"

File: pydevd.py [hunk 2]
```diff
@@ -542,6 +564,8 @@ def process_internal_commands(self):
     def set_tracing_for_untraced_contexts(self, ignore_frame=None, overwrite_prev_trace=False):
         # Enable the tracing for existing threads (because there may be frames being executed that
         # are currently untraced).
+        if self.frame_eval_func is not None:
```
Comment (Elizaveta239) [inline pydevd.py]: Fixed

File: pydevd.py [hunk 3]
```diff
@@ -957,11 +976,17 @@ def run(self, file, globals=None, locals=None, is_module=False, set_trace=True):
                 # The file being run must be in the pythonpath (even if it was not before)
                 sys.path.insert(0, os.path.split(file)[0])
 
-            self.prepare_to_run()
-
             while not self.ready_to_run:
                 time.sleep(0.1)  # busy wait until we receive run command
 
+            if self.break_on_caught_exceptions or (self.plugin and self.plugin.has_exception_breaks()) or self.signature_factory:
+                # disable frame evaluation if there are exception breakpoints with 'On raise' activation policy
```
Comment (Elizaveta239) [inline pydevd.py]: Yes, you're right. I've added checks for adding exception breakpoints during run.
But I didn't get your phrase about `trace_exception`. What do you mean here? Frame evaluation debugger can't handle exceptions.

File: _pydevd_frame_eval/pydevd_frame_tracing.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+dummy_tracing_holder = DummyTracingHolder()
+
+
+def update_globals_dict(globals_dict):
+    new_globals = {'_pydev_stop_at_break': _pydev_stop_at_break}
+    globals_dict.update(new_globals)
+
+
+def handle_breakpoint(frame, thread, global_debugger, breakpoint):
+    # ok, hit breakpoint, now, we have to discover if it is a conditional breakpoint
```
Comment (Elizaveta239) [inline _pydevd_frame_eval/pydevd_frame_tracing.py]: Done, moved these to separate methods in `pydevd_frame.py`

File: _pydevd_frame_eval/pydevd_frame_eval_cython_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        except:
+            # In Jython this call fails, but this is Ok, we don't support Jython for speedups anyways.
+            raise ImportError
+        plat = '32'
+        if is_64bits:
+            plat = '64'
+
+        # We also accept things as:
+        #
+        # _pydevd_bundle.pydevd_cython_win32_27_32
```
Comment (Elizaveta239) [inline _pydevd_frame_eval/pydevd_frame_eval_cython_wrapper.py]: Fixed

File: _pydevd_bundle/pydevd_comm.py [hunk 6]
```diff
@@ -1437,3 +1437,31 @@ def pydevd_find_thread_by_id(thread_id):
...
+            try:
+                additional_info = t.additional_info
+            except AttributeError:
+                pass  # that's ok, no info currently set
+            if additional_info is None:
+                continue
+
+            for frame in additional_info.iter_frames(t):
+                main_debugger.set_trace_for_frame_and_parents(frame, overwrite_prev_trace=True)
+            main_debugger.set_use_code_extra(False)
```
Comment (Elizaveta239) [inline _pydevd_bundle/pydevd_comm.py]: Yes, you're right. I've renamed it to "enable_cache_frames_without_breaks".

File: _pydevd_bundle/pydevd_frame.py [hunk 7]
```diff
@@ -674,16 +698,13 @@ def trace_dispatch(self, frame, event, arg):
...
-                    # while stepping when execution is finished
-                    if is_return:
-                        f_code = getattr(frame.f_back, 'f_code', None)
-                        if f_code is not None:
-                            back_filename = os.path.basename(f_code.co_filename)
-                            file_type = get_file_type(back_filename)
-                            if file_type == PYDEV_FILE:
-                                stop = False
+                if step_cmd != -1 and is_return and IS_PY3K and hasattr(frame, "f_back"):
+                    f_code = getattr(frame.f_back, 'f_code', None)
```
Comment (Elizaveta239) [inline _pydevd_bundle/pydevd_frame.py]: I removed `stop` accidentally, you're right. But we need to check whether a `frame` object has a `f_back`. It isn't always true, so we need to check it.

**Comment by Elizaveta239** [comment]
Hi Fabio! I've made fixes according to your comments. 
It looks like I missed something from your comment about adding exception breakpoints during running. Could you please review the latest changes and let me know if it isn't what you meant?
Also your idea about adding tests in order to check switching between frame evaluation and tracing debugger was very good! I've added tests in the latest commit.

**Comment by achuchev** [comment]
@Elizaveta239 Good job. 
@fabioz Do you plan to merge this PR?

**Comment by fabioz** [comment]
Hi @achuchev yes, unfortunately I've run out of time with some other things and had to put the review on hold. I'll get into a 2-week vacation next week and my plan is reviewing this (and the other pending pull requests) the next week after I come back from vacations (in the week of August 7th).

**Comment by achuchev** [comment]
@fabioz Thank you, have a nice holiday!

**Comment by Elizaveta239** [comment]
@fabioz It's good, that you haven't merged it yet, because we have a few fixes in frame evaluation debugger which we made after creation of this pull request. I'll add them here on the next week.

**Comment by fabioz** [comment]
I've just merged this to `master` (doesn't appear as merged because it was targeting another branch).

Thanks for the patch Elizaveta ;)

**Pull request closed without merging**